### PR TITLE
vim-patch:8.1.{1631,1682,1899}: sign improvements

### PR DESCRIPTION
**Pull request by lewis6991**
### vim-patch:8.1.1631: displaying signs is inefficient
Original PR: https://github.com/vim/vim/pull/4586

> When displaying a sign, multiple calls are made to get information
related to a sign. Each call walks through all the signs placed
in the buffer. Add a function to get all the sign attributes in a single
call and use it.

**Note:** This patch is incompatibilites with Neovim's dynamic signcolumn. I have adapted the code as best I can to achieve the same effect of the patch by calcuating the sign attributes once for all signs on the line.

### vim-patch:8.1.1682: placing a larger number of signs is too slow
Original PR: https://github.com/vim/vim/pull/4636

> vim-patch:8.1.1682: placing a larger number of signs is too slow
> Add support for:
> - defining multiple signs using sign_define()
> - undefining multiple signs using sign_undefine()
> - placing multiple signs using sign_placelist()
> - removing multiple signs using sign_unplacelist()

### vim-patch:8.1.1899: sign_place() does not work as documented 
Original PR: https://github.com/vim/vim/pull/4848

> The help text for sign_place() states that the lnum argument accepts values
that are supported by line(). Modify the code to make it so.

### CODE REVIEW

File: src/nvim/testdir/otherName [hunk 1]
```diff
@@ -0,0 +1,2 @@
+some
```
Comment (janlazo) [inline src/nvim/testdir/otherName:1]: Delete

File: src/nvim/testdir/test_result.log [hunk 2]
```diff
@@ -0,0 +1,147 @@
+
```
Comment (janlazo) [inline src/nvim/testdir/test_result.log:1]: Delete

File: src/nvim/version.c [hunk 3]
```diff
@@ -67,6 +67,7 @@ NULL
 
 // clang-format off
 static const int included_patches[] = {
+  1899,
```
Comment (janlazo) [inline src/nvim/version.c:70]: Revert.

These patch numbers target Vim 8.0. Neovim has patches across multiple Vim versions so `included_patches` updates are automated.

**Review by janlazo — CHANGES_REQUESTED** [review]

File: src/nvim/sign.c [hunk 4]
```diff
@@ -974,7 +977,7 @@ int sign_place(
         has_text_or_icon);
   } else {
     // ":sign place {id} file={fname}": change sign type
```
Comment (janlazo) [inline src/nvim/sign.c:979]: Vim changed this comment.

File: src/nvim/sign.c [hunk 5]
```diff
@@ -1860,3 +1863,267 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
...
+  char *numhl = NULL;
+  int retval = -1;
+
+  if (name_arg == NULL) {
+    if (dict == NULL) {
+      return -1;
+    }
+    name = tv_dict_get_string(dict, "name", true);
+  } else {
+    name = (char *)vim_strsave((char_u *)name_arg);
```
Comment (janlazo) [inline src/nvim/sign.c:1884]: Use `xstrdup` directly.

File: src/nvim/testdir/test_signs.vim [hunk 6]
```diff
@@ -1143,7 +1153,7 @@ func Test_sign_id_autogen()
   call writefile(repeat(["Sun is shining"], 30), "Xsign")
   edit Xsign
 
-  call assert_equal(1, sign_place(0, '', 'sign1', 'Xsign',
+  call assert_equal(2, sign_place(0, '', 'sign1', 'Xsign',
```
Comment (janlazo) [inline src/nvim/testdir/test_signs.vim:1156]: Bug?

File: src/nvim/testdir/test_signs.vim [hunk 7]
```diff
@@ -1159,8 +1169,8 @@ func Test_sign_id_autogen()
   call assert_equal(2, sign_place(0, 'g1', 'sign1', 'Xsign',
 	      \ {'lnum' : 12}))
   call assert_equal(0, sign_unplace('g1', {'id' : 1}))
-  call assert_equal(10,
-	      \ sign_getplaced('Xsign', {'id' : 1})[0].signs[0].lnum)
+  call assert_equal(12,
```
Comment (janlazo) [inline src/nvim/testdir/test_signs.vim:1172]: Bug?

File: src/nvim/testdir/test_signs.vim [hunk 8]
```diff
@@ -1869,3 +1879,121 @@ func Test_sign_numcol()
...
+	      \ 'group' : 'g2', 'priority' : 100},
+	      \ {'id' : 1, 'name' : 'sign1', 'lnum' : 11,
+	      \ 'group' : 'g1', 'priority' : 50},
+	      \ {'id' : 3, 'name' : 'sign3', 'lnum' : 11,
+	      \ 'group' : '', 'priority' : 10}], s[0].signs)
+
+  call sign_unplace('*')
+
+  " Place multiple signs at once with auto-generated sign identifier
+  call assert_equal([1, 1, 1], sign_placelist([
```
Comment (janlazo) [inline src/nvim/testdir/test_signs.vim:1927]: [1,1,5]

**Review by janlazo — CHANGES_REQUESTED** [review]
There are extra changes in test_signs.vim which may suggest missing patches or undocumented difference between Vim and Neovim.

Please fix the minor issues I identified. The rest looks fine.

File: src/nvim/testdir/test_signs.vim (hunk 6 above)
Comment (lewis6991) [reply src/nvim/testdir/test_signs.vim:1156]: Still need to investigate whether this is ok. As it is only an id I don't think it needs to be identical across versions of the code, it just need to be unique.

File: src/nvim/testdir/test_signs.vim (hunk 7 above)
Comment (lewis6991) [reply src/nvim/testdir/test_signs.vim:1172]: As above

File: src/nvim/testdir/test_signs.vim (hunk 8 above)
Comment (lewis6991) [reply src/nvim/testdir/test_signs.vim:1927]: Again these are only id's so I don't think we need to have the same results as vim. The test placeing signs for three different groups so it's ok that the signs all have the same id.

**Comment by lewis6991** [comment]
Status update on this.

I ended up also adding vim patch 8.1.1631 which provides performance improvements by calculating sign attributes ahead of time instead of doing multiple of lookups for each, which looking at it is really inefificent. Because of neovim's multi-signcolumn feature I had to significantly rework this patch so we calculate upto 9 sets of sign attrs instead of 1.

The test discrepency in the last review was due to test ordering. I omitted part of the original patch which added `aaa` into a test name to force it to run before other tests. After adding this, the test changes could be reverted to match vim.

Other than that I think I've addressed all the previous review comments but a may have missed one.

Thanks.

**Review by janlazo — APPROVED** [review]
LGTM

**Pull request merged**